privacyidea/lib/policymatch.py

```diff
@@ -0,0 +1,175 @@
+            username = None
+            realm = None
+            resolver = None
+        else:
+            raise MatchingError("Invalid user")
+        return cls(g, name=None, scope=scope, realm=realm, active=True,
+                   resolver=resolver, user=username, user_object=user_object,
+                   client=g.client_ip, action=action, adminrealm=None, time=None,
+                   sort_by_priority=True)
+
+    @classmethod
+    def admin(cls, g, action, realm):
+        """
+        Matching admin policies with an action and, optionally, a realm.
+        Assumes that the currently logged-in user is an admin, and throws an error otherwise.
+        Policies will be matched against the admin's username and adminrealm,
+        and optionally also the provided realm.
+        :param g: ``flask.g`` object
+        :param action: the policy action
+        :param realm: the realm against which policies should be matched. Can be None.
+        :rtype: ``Match``
+        """
+        username = g.logged_in_user["username"]
+        adminrealm = g.logged_in_user["realm"]
+        from privacyidea.lib.auth import ROLE
+        if g.logged_in_user["role"] != ROLE.ADMIN:
+            raise MatchingError("Policies with scope ADMIN can only be retrieved by admins")
+        return cls(g, name=None, scope=SCOPE.ADMIN, realm=realm, active=True,
+                   resolver=None, user=username, user_object=None,
+                   client=g.client_ip, action=action, adminrealm=adminrealm, time=None,
+                   sort_by_priority=True)
+
+    @classmethod
+    def admin_or_user(cls, g, action, realm):
+        """
+        Depending on the role of the currently logged-in user, match either scope=ADMIN or scope=USER policies.
+        If the currently logged-in user is an admin, match policies against the username, adminrealm
+        and the given realm. If the currently logged-in user is a user, match policies against the username
+        and the given realm.
+        :param g: ``flask.g`` object
+        :param action: the policy action
+        :param realm: the given realm, must not be None
+        :rtype: ``Match``
+        """
+        from privacyidea.lib.auth import ROLE
+        if g.logged_in_user["role"] == ROLE.ADMIN:
+            scope = SCOPE.ADMIN
+            username = g.logged_in_user["username"]
+            adminrealm = g.logged_in_user["realm"]
+        elif g.logged_in_user["role"] == ROLE.USER:
+            scope = SCOPE.USER
+            username = g.logged_in_user["username"]
+            adminrealm = None
+            # TODO: resolver?
+        else:
+            # TODO: API key?
+            raise MatchingError("Unknown role")
+        return cls(g, name=None, scope=scope, realm=realm, active=True,
+                   resolver=None, user=username, user_object=None,
+                   client=g.client_ip, action=action, adminrealm=adminrealm, time=None,
+                   sort_by_priority=True)
```

Comment:
Please add a further constructor, that reflects the interface of ``match_policies`` so that it can also be used e.g. in ``check_base_action`` called "generic".

Reply:
Sounds good! However, I'd like to do that in a later PR in order to not make this PR even bigger :)